src/openms/source/ANALYSIS/TOPDOWN/DeconvolutedSpectrum.cpp

```diff
@@ -0,0 +1,654 @@
+               "IsotopeCosine\tMassSNR\tMaxQScoreCharge\tMaxQScoreMzStart\tMaxQScoreMzEnd\tQScore\n";
+      }
+    }
+  }
+
+  void DeconvolutedSpectrum::writeAttCsvHeader(std::fstream &fs)
+  {
+    fs
+        << "ScanNumber,RetentionTime,PrecursorScanNumber,Charge,ChargeSNR,PeakIntensity,EnvIntensity,EnvIsotopeCosine,PeakMz,"
+           "MonoMass,MassSNR,IsotopeCosine,ChargeIntensityCosine,MassIntensity,QScore,Class\n";
+  }
+
+  void DeconvolutedSpectrum::writeThermoInclusionHeader(std::fstream &fs)
+  {
+    fs << "Compound,Formula,Adduct,m/z,z,t start (min),t stop (min),Isolation Window (m/z),Normalized AGC Target (%)\n";
+  }
+
+
+  void DeconvolutedSpectrum::clearPeakGroupsChargeInfo()
+  {
+    for (auto &pg: peakGroups)
+    {
+      pg.clearChargeInfo();
+    }
+  }
+
+  void DeconvolutedSpectrum::writeAttCsv(std::fstream &fs,
+                                         int msLevel,
+                                         double qScoreThreshold = -1000,
+                                         int numMaxMS2 = -1)
+  {
+
+    if (msLevel > 1)
+    {
+      if (precursorPeakGroup != nullptr)
+      {
+        fs << scanNumber << "," << precursorPeakGroup->spec->getRT() << "," << precursorPeakGroup->scanNumber << ","
+           << precursorPeak.getCharge()
+           << "," << log10(precursorPeakGroup->perChargeInfo[precursorPeak.getCharge()][0] + 1e-3) << ","
+           << log10(precursorPeak.getIntensity() + 1)
+           << "," << log10(precursorPeakGroup->perChargeInfo[precursorPeak.getCharge()][2] + 1)
+           << "," << precursorPeakGroup->perChargeInfo[precursorPeak.getCharge()][1]
+           << "," << precursorPeak.getMZ()
+           << "," << precursorPeakGroup->monoisotopicMass << "," << log10(precursorPeakGroup->totalSNR + 1e-3) << ","
+           << precursorPeakGroup->isotopeCosineScore
+           << "," << precursorPeakGroup->chargeCosineScore << "," << log10(precursorPeakGroup->intensity + 1)
+           << "," << precursorPeakGroup->qScore
+           << ",f\n";
+      }
+      else
+      {
+        fs << scanNumber << "," << spec->getRT() << "," << 0 << "," << 0
+           << ",?,"
+           << log10(precursorPeak.getIntensity() + 1)
+           << ",?,?," << (precursorPeak.getMZ())
+           << "," << 0 << ",?,?,?,?"
+           << "," << precursorPeakGroup->qScore
+           << ",f\n";
+      }
+    }
+    else
+    {
+      //double scoreThreshold = 0;
+      std::vector<double> scores;
+
+      if (numMaxMS2 > 0 && peakGroups.size() > (Size) numMaxMS2)// max peak count for TopPic
+      {
+        scores.reserve(peakGroups.size());
+        for (auto &pg : peakGroups)
+        {
+          scores.push_back(pg.qScore);
+        }
+        std::sort(scores.begin(), scores.end());
+        qScoreThreshold = std::max(qScoreThreshold, scores[scores.size() - numMaxMS2]);
+        std::vector<double>().swap(scores);
```

Comment:
I think the combination of reserve and the swap with empty idiom might be slower as it could lead to reallocations. maybe just scores.clear() ? it does not free the memory (fast)

Reply:
It does not free... as far as I know  

https://stackoverflow.com/questions/35514909/how-to-clear-vector-in-c-from-memory

Reply:
your vector scores is declared outside of the loop. If you do the swap, the memory is freed and a fresh allocation is needed in the next iteration.
if you do a scores.clear() it will set the vector size to 0 but will keep the memory internally reserved (actually it will keep the old data in there without deleting it ) and no reallocation will occur in the next iteration (except if the internal capacity is exceeded). This can speed up things quite a bit compared to doing the swap.